add extrudeHelical

Loosetooth opened the pull request:
Adds the `extrudeHelical` functionality as discussed in #1162

This supposes that the `geom2` that is to be extruded is laying fully in the positive x plane. Otherwise you get issues with inside-out polygons, as discussed in #1012

It allows for the generating of a helix, tapered helix (`endRadiusOffset !== 0`) and spiral (`pitch === 0, endRadiusOffset !== 0`)

Simple example:
```js
const jscad = require("@jscad/modeling")
const { circle } = jscad.primitives
const { extrudeHelical } = jscad.extrusions

const main = () => {
    return extrudeHelical(
            {
                angle: Math.PI * 4,
                pitch: 10,
                endRadiusOffset: -5,
                segments: 64
            },
            circle({ size: 3, center: [10, 0] })
        )
}

module.exports = { main }
```

Resulting in a tapered helix:
![image](https://user-images.githubusercontent.com/66420755/200585281-bd7a4999-ac15-44cf-9a81-a62256fc95b2.png)

I will need some extra guidance with adding the documentation, as I could not immediately find where I should add such documentation. For example, there's the [docs](https://openjscad.xyz/docs/) website, and the [user guide](https://openjscad.xyz/dokuwiki/doku.php). I'm not sure which one is supposed to be edited, and how I do that. Thanks for the help.

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?

z3dev commented:
@Loosetooth WOW! That's was super fast! Nice job!

I'll look over the code on the weekend. :)

The API documentation is generated from the source code. You can generate a local copy by...

```
nom run docs
open docs/index.html
```

Loosetooth commented:
Thanks @z3dev , looking forward to your feedback.

z3dev commented on packages/modeling/src/operations/extrusions/extrudeHelical.js:
```diff
@@ -0,0 +1,81 @@
+ *
+ * @param {Object} options - options for extrusion
+ * @param {Number} [options.angle=TAU] - angle of the extrusion (RADIANS) positive for right-hand rotation, negative for left-hand
+ * @param {Number} [options.pitch=10] - elevation gain for each turn
```
I like both 'pitch' and 'height'. I can see some users wanting to specify the pitch for each turn, but some users may want to set a final 'height'. I'm wondering if we could support both of these.

z3dev commented on packages/modeling/src/operations/extrusions/extrudeHelical.js:
```diff
@@ -0,0 +1,81 @@
+ * @param {Object} options - options for extrusion
+ * @param {Number} [options.angle=TAU] - angle of the extrusion (RADIANS) positive for right-hand rotation, negative for left-hand
+ * @param {Number} [options.pitch=10] - elevation gain for each turn
+ * @param {Number} [options.endRadiusOffset=0] - offset the final radius of the extrusion, allowing for tapered helix, and or spiral
```
I hate names... endRadiusOffset seems a little long. Does something like 'endOffset' fit?

z3dev commented on packages/modeling/src/operations/extrusions/extrudeHelical.js:
```diff
@@ -0,0 +1,81 @@
+ * Perform a helical extrude of the geometry, using the given options.
+ *
+ * @param {Object} options - options for extrusion
+ * @param {Number} [options.angle=TAU] - angle of the extrusion (RADIANS) positive for right-hand rotation, negative for left-hand
```
@Loosetooth is it possible to support 'startAngle'? both extrudeRotate and torus have this option.

z3dev reviewed (changes requested):
@Loosetooth code looks good but please adjust the formatting.

Also, lots of questions / thoughts about the parameters. Please provide some nice ideas. ;)

z3dev commented:
@Loosetooth it seems that full rotation is not being realized. here are some screen shots. i suspect that you will have a lot of fun with these 'rotations'. the extrudeRotate function has some logic to calculate rotations, which might help.

z3dev commented:
![issue1](https://user-images.githubusercontent.com/8969165/201504305-cdbaf379-f033-47df-b5cb-2408a16767bf.png)
![issue2](https://user-images.githubusercontent.com/8969165/201504307-e9d4b9ef-9d45-4c37-923f-c2eb7cb3ad2e.png)

Loosetooth commented:
Hi @z3dev , thanks for the feedback.

Implementing the `startAngle` parameter shouldn't be a problem.
Full rotations should indeed be fixed, good catch. It's doing one step too little.

> I like both 'pitch' and 'height'. I can see some users wanting to specify the pitch for each turn, but some users may want to set a final 'height'. I'm wondering if we could support both of these.

Sure, this is also a parameter that some implementations in other CAD software accept. The only reason I did not implement it immediately was that it's slightly ambiguous. How do we measure the height of the helix? Do we measure from the center of the start sketch, until the center of the final extruded sketch? Or do we measure from the "bottom" of the start sketch, until the top of the final sketch?

The most straightforward implementation would be the first option, then we automatically calculate the pitch in this manner:
```js
const pitch = height / (angle / TAU)
```
In this case the total height of the helix would however be larger than the provided `height` parameter, which is perhaps confusing to users:
```js
const totalHelixHeight = height + sketchHeight
```

What do you think about this solution?

>  hate names... endRadiusOffset seems a little long. Does something like 'endOffset' fit?

Agree that the name is too long. But I'm not sure if `endOffset` would be a good choice as it is not as descriptive. Offset can be interpreted in multiple ways.  I have no other ideas right now for a name though :D If you have any other ideas please share, if we don't find anything else I don't mind to name it endOffset.

Loosetooth commented:
Please have a look @z3dev 

Implemented all the requested changes.
The height parameter, when specified refers to the total height of the extrusion path, not the height of the extruded solid, so should be clear to users.

Renamed `endRadiusOffset` to `endOffset`, so the basic example should now be:
```js
const jscad = require("@jscad/modeling")
const { circle } = jscad.primitives
const { extrudeHelical } = jscad.extrusions

const main = () => {
    return extrudeHelical(
            {
                angle: Math.PI * 4,
                pitch: 10,
                endOffset: -5,
                segments: 64
            },
            circle({ size: 3, center: [10, 0] })
        )
}

module.exports = { main }
```

z3dev commented on packages/modeling/src/operations/extrusions/extrudeHelical.js:
```diff
@@ -0,0 +1,95 @@
+    pitch = options.pitch ? options.pitch : defaults.pitch
+  }
+
+  if (segments < 2) throw new Error('segments must be greater than 1')
```
hmm... i don't see valid results coming from 2 segments. maybe minimum is 3.

Loosetooth replied:
Thinking about this, it needs at least 3 segments for each complete revolution, in order to get a valid geometry. Even if it's a partial revolution, so we round up.

To see this, we can find an angle $\alpha$ for every number of segments $n \in \mathbb{Z}^{+}$ so that

$$
{\alpha \over n} = \pi
\Leftrightarrow
\alpha = \pi n
$$

Which means that for every number of segments we can find an angle that will project all the slices in the same plane, resulting in incorrect (flat) geometry. Requiring that the number of segments is larger than any arbitrary value can never solve this problem.

z3dev reviewed (changes requested):
@Loosetooth really close now. i'm looking forward to using this. nice.

just a few more suggestions.

Loosetooth commented on packages/modeling/src/operations/extrusions/extrudeHelical.d.ts:
```diff
@@ -0,0 +1,12 @@
+export interface ExtrudeHelicalOptions {
+  angle?: number
+  pitch?: number
+  endRadiusOffset?: number
```
Fixed in 86ed105

Loosetooth commented on packages/modeling/src/operations/extrusions/extrudeHelical.js:
```diff
@@ -0,0 +1,95 @@
+      mat4.fromXRotation(mat4.create(), -TAU / 4) // putting TAU/4 here creates inside-out polygon
+    )
+
+    const matrix = mat4.create()
```
Added in 25f3b7b, please let me know is this is what you meant.

z3dev commented:
@Loosetooth try this... i think you will have to flip the slice depending on the direction of the rotation.

```
  const s1 = extrudeHelical(
            {
                angle: constants.TAU * -1.25,
                startAngle: constants.TAU * 0.2,
                pitch: -3,
                endRadiusOffset: 0,
                segments: 64
            },
            // circle({ size: 1, center: [10, 0] })
            ellipse({ radius: [5, 1], center: [5, 0] })
        )
```

Loosetooth commented:
> @Loosetooth try this... i think you will have to flip the slice depending on the direction of the rotation.
> 
> ```
>   const s1 = extrudeHelical(
>             {
>                 angle: constants.TAU * -1.25,
>                 startAngle: constants.TAU * 0.2,
>                 pitch: -3,
>                 endRadiusOffset: 0,
>                 segments: 64
>             },
>             // circle({ size: 1, center: [10, 0] })
>             ellipse({ radius: [5, 1], center: [5, 0] })
>         )
> ```

True, for negative angles, the slice has to be flipped in the other direction.

This should now work without inside out polygons:
```js
const jscad = require("@jscad/modeling")
const { ellipse } = jscad.primitives
const { extrudeHelical } = jscad.extrusions


const main = () => {
    return extrudeHelical(
        {
            angle: Math.PI * 2 * -1.25,
            startAngle: Math.PI * 2 * 0.2,
            pitch: -3,
            endOffset: 0,
            segments: 64
        },
        // circle({ size: 1, center: [10, 0] })
        ellipse({ radius: [5, 1], center: [5, 0] })
    )
}

module.exports = { main }
```

It will still give bad results when the sketch is not in the positive x plane though. I should fix that :)

Loosetooth commented:
> @z3dev , is there an easy way to invert the 'direction' of `geom2` shapes? What I mean is that in one orientation (clockwise?) extruding the shape results in a normal geom3 polygon. When the orientation is opposite (anti clockwise?) the extrusion results in an inverted polygon.
> 
> If we can easily invert this orientation, we can detect if the shape is in the negative x plane. Then we invert the orientation, instead of changing the rotation transformations, to get a correctly extruded polygon.

Found it :)

Geom2's with all points fully negative X (zero is counted as positive) can now also be helically extruded.

z3dev commented:
@Loosetooth this is working really well now. super thanks.

However, while playing around with some of the different types of extrusions, I created this combination.

  ```
const s1 = extrudeHelical(
            {
                angle: constants.TAU * 10.25,
                startAngle: constants.TAU * 0.2,
                height: 20,
                endRadiusOffset: 0,
                segments: 64 * 10.25  // STRANGE??
            },
            ellipse({ radius: [2, 1], center: [5, 0] })
        )
```

It seems that the user has to calculate the number of segments for the full rotation. Not wrong, but I wanted to point this out for discussion. I was wondering if a user would rather specify the segments PER rotation instead.

Thoughts?

Loosetooth commented:
Sure, let's do segments per rotation. This would also simplify some error handling in the code.

How would you like to name this parameter? I suppose `segmentsPerRotation` is too long for your liking. If we keep it `segments`, we might confuse users because it means something else in all other extrusion functions.

z3dev commented:
> Sure, let's do segments per rotation. This would also simplify some error handling in the code.
> 
> How would you like to name this parameter? I suppose `segmentsPerRotation` is too long for your liking. If we keep it `segments`, we might confuse users because it means something else in all other extrusion functions.

Ha ha ha! SegmentsPerRotation is fine.

z3dev commented on packages/modeling/src/operations/extrusions/extrudeHelical.js:
```diff
@@ -13,7 +13,7 @@ const geom2 = require('../../geometries/geom2')
  * @param {Number} [options.height] - total height of the helix path. Ignored if pitch is set.
  * @param {Number} [options.endOffset=0] - offset the final radius of the extrusion, allowing for tapered helix, and or spiral
- * @param {Number} [options.segments=32] - number of segments of the extrusion
+ * @param {Number} [options.segmentsPerRotation=32] - number of segments of the extrusion
```
... number of segments per rotation of the extrusion

Loosetooth replied:
Is already in, looking at old commit. Still had some commits after my comment, apologies.

z3dev reviewed (changes requested):
@loosetooth just a small comment change then this is ready.

Loosetooth commented on packages/modeling/src/operations/extrusions/extrudeHelical.js:
```diff
@@ -0,0 +1,104 @@
+  const minNumberOfSegments = Math.ceil(Math.abs(angle) / TAU) * 3
+
+  if (segments < minNumberOfSegments)
+    throw new Error(`For a rotation of ${angle - startAngle} radians there need to be at least ${minNumberOfSegments} segments.
```
This is already changed, you're looking at an old commit.

Loosetooth commented on packages/modeling/src/operations/extrusions/extrudeHelical.js:
```diff
@@ -0,0 +1,104 @@
+
+
+  // define transform matrix variables for performance increase
+  let step1
```
Thanks for the clarification, added in [f5eb972](https://github.com/jscad/OpenJSCAD.org/pull/1164/commits/f5eb9729c223ed732bc12b302a916265d04bc600)

z3dev reviewed (approved):
@Loosetooth super. This looks good now.

I don't have any more suggestions or issues. Let's see if @platypii or @hrgdavor  have any suggestions.

hrgdavor reviewed (approved):
A nice addition. Thanks for the contribution. I have been following the conversation and changes. @z3dev is much better at noticing details, and to me it looks ready for merge :)

z3dev commented:
@Loosetooth Thanks again for the nice addition. Sorry that it took extra time but the final outcome was really nice.
